File: Next_Permutation.java
import java.util.*;

public class Solution {
    public void nextPermutation(int[] array) {
        int i, j;
        // Find the largest index k such that a[k] < a[k + 1]. If no such index
        // exists, the permutation is the last permutation.
        for (i = array.length - 2; i >= 0; i--) {
            if (array[i] < array[i + 1])
                break;
        }
        if (i < 0) {
            // System.out.println("End");
            Arrays.sort(array);
            return;
        }
        
        // Find the largest index l such that a[k] < a[l]. Since k + 1 is such
        // an index, l is well defined and satisfies k < l.
        for (j = array.length - 1; j > i; j--) {
            if (array[j] > array[i])
                break;
        }
        
        // Swap a[k] with a[l].
        swap(array, i++, j);
        
        // Reverse the sequence from a[k + 1] up to and including the final
        // element a[n].
        for (j = array.length - 1; j > i; i++, j--) {
            swap(array, i, j);
        }
    }
    
    public void swap(int[] array, int x, int y) {
        array[x] ^= array[y];
        array[y] ^= array[x];
        array[x] ^= array[y];
    }
}

File: Swift/next-permutation.swift
/**
 * https://leetcode.com/problems/next-permutation/
 * 
 * 
 */ 
/**
 Approach 2: Single Pass Approach

 Algorithm

 First, we observe that for any given sequence that is in descending order, no next larger permutation is possible. For example, no next permutation is possible for the following array:

 [9, 5, 4, 3, 1]
 
 We need to find the first pair of two successive numbers a[i] and a[i−1], from the right, which satisfy
 a[i]>a[i−1]. Now, no rearrangements to the right of a[i−1] can create a larger permutation since that subarray consists of numbers in descending order. Thus, we need to rearrange the numbers to the right of a[i−1] including itself.

 Now, what kind of rearrangement will produce the next larger number? We want to create the permutation just larger than the current one. Therefore, we need to replace the number a[i−1] with the number which is just larger than itself among the numbers lying to its right section, say a[j].

Next Permutation

 We swap the numbers
 a[i−1] and a[j]. We now have the correct number at index i−1. But still the current permutation isn't the permutation that we are looking for. We need the smallest permutation that can be formed by using the numbers only to the right of a[i−1]. Therefore, we need to place those numbers in ascending order to get their smallest permutation.

 But, recall that while scanning the numbers from the right, we simply kept decrementing the index until we found the pair a[i] and a[i−1] where, a[i]>a[i−1]. Thus, all numbers to the right of a[i−1] were already sorted in descending order. Furthermore, swapping a[i−1] and a[j] didn't change that order. Therefore, we simply need to reverse the numbers following a[i−1] to get the next smallest lexicographic permutation.
*/

class Solution {
    func nextPermutation(_ nums: inout [Int]) {
        var i = nums.count - 2
        while i >= 0 {
            if nums[i] < nums[i + 1] { break }
            i -= 1
        }
        if i < 0 {
            // Last permutation.
            nums.sort()
            return
        }
        
        var j = nums.count - 1
        while j > i {
            if nums[j] > nums[i] { break }
            j -= 1
        }
        nums.swapAt(i, j)
        i += 1
        j = nums.count - 1
        while i < j {
            nums.swapAt(i, j)
            i += 1
            j -= 1
        }
    }
}
